### PR TITLE
Remove setAndCheckCallbackStatus* from ParserCallbacks.

### PR DESCRIPTION
**Pull request by bencebeky**
Remove setAndCheckCallbackStatus() and setAndCheckCallbackStatusOr()
methods from ParserCallbacks interface.  ParserCallbacks currently
requires the Parser implementation to call the respective method on the
return value of every single non-void method, which is fragile.

The only ParserCallbacks implementation is ConnectionImpl and its
derived classes, and setAndCheckCallbackStatus() and
setAndCheckCallbackStatusOr() really are implementation details therein,
therefore they belong to ConnectionImpl.

Overrides of ParserCallbacks methods in ConnectionImpl often have
multiple returns, so the cleanest solution is to rename the current
non-void overrides *Impl() for the ones implemented in ConnectionImpl
and *Base() (following three existing examples) for onUrlBase() and
onStatusBase() implemented in ServerConnectionImpl and
ClientConnectionImpl, and make the ParserCallbacks implementation call
into these *Impl() and *Base() methods, making sure that
setAndCheckCallbackStatus() or setAndCheckCallbackStatusOr() is then
called with the return value.

This way all non-void ParserCallbacks methods can return ParseStatus.

This simplifies ParserCallbacks and LegacyParserImpl, as well as the
future other Parser implementation, at the cost of complicating
ConnectionImpl.

Also move onStatus() above onHeaderField() and onHeaderValue(), because
it is called earlier during parsing.

Also move pure virtual onHeadersCompleteBase() and
onMessageCompleteBase() declarations from the private section to the
protected one in ConnectionImpl, because even though private methods can
be overriden (but not called) in derived classes, this is not widely
known and is confusing.  This is consistent with the existing
onMessageBeginningBase() declaration.

Tracking issue: https://github.com/envoyproxy/envoy/issues/21245

Signed-off-by: Bence Béky <bnc@google.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/http/http1/parser.h [hunk 1]
```diff
@@ -46,47 +46,47 @@ class ParserCallbacks {
...
    * @return Status representing success or failure.
    */
-  virtual Status onStatus(const char* data, size_t length) PURE;
+  virtual ParserStatus onHeaderValue(const char* data, size_t length) PURE;
 
   /**
    * Called when headers are complete. A base routine happens first then a virtual dispatch is
    * invoked. Note that this only applies to headers and NOT trailers. End of
    * trailers are signaled via onMessageCompleteBase().
    * @return An error status or a ParserStatus.
```
Comment (KBaichoo) [inline source/common/http/http1/parser.h:87]: update return since only ParserStatus returned

File: source/common/http/http1/codec_impl.h [hunk 2]
```diff
@@ -344,19 +347,26 @@ class ConnectionImpl : public virtual Connection,
...
+  ParserStatus onHeaderValue(const char* data, size_t length) override;
+  ParserStatus onHeadersComplete() override;
   void bufferBody(const char* data, size_t length) override;
-  StatusOr<ParserStatus> onMessageComplete() override;
-  ParserStatus setAndCheckCallbackStatus(Http::Status&& status) override;
-  ParserStatus setAndCheckCallbackStatusOr(Envoy::StatusOr<ParserStatus>&& statusor) override;
-
-  // Connection specific callback methods.
-  virtual Envoy::StatusOr<ParserStatus> onHeadersCompleteBase() PURE;
-  virtual ParserStatus onMessageCompleteBase() PURE;
```
Comment (KBaichoo) [inline source/common/http/http1/codec_impl.h:357]: Re: moving the method from private -> protected, it weakens encapsulation as there's now a larger swath of potential callers. I think this is a common enough pattern in C++. 

From Google C++ styleguide: Limit the use of protected to those member functions that might need to be accessed from subclasses. in https://google.github.io/styleguide/cppguide.html#Inheritance

Perhaps onMessageBeginningBase should become consistent with these

File: source/common/http/http1/parser.h (hunk 1 above)
Comment (bencebeky) [reply source/common/http/http1/parser.h]: Done.  I'm also updating ParserStatus description.

File: source/common/http/http1/parser.h [hunk 3]
```diff
@@ -99,15 +99,12 @@ class ParserCallbacks {
    * Called when the HTTP message has completed parsing.
    * @return An error status or a ParserStatus.
```
[the remark this answers is not in the document]
Comment (bencebeky) [reply source/common/http/http1/parser.h]: Done.

**Comment by bencebeky** [comment]
Kevin: PTAL.  Thank you.

**Review by KBaichoo — APPROVED** [review]
lgtm pending CI

**Comment by KBaichoo** [comment]
/assign @mattklein123 

For final pass

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**